spacy/tokens/doc.pyx

```diff
@@ -1743,6 +1743,124 @@ cdef class Doc:
+        ps_ variables relate to searches starting at the beginning of the word
+        ss_ variables relate to searches starting at the end of the word
+        
+        cs: if *False*, hashes are generated based on the lower-case version of each token.
+        p_lengths: an array of single-byte values specifying the lengths of prefixes to be hashed in ascending order. 
+            For example, if *p_lengths==[2, 3]*, the prefixes hashed for "spaCy" would be "sp" and "spa".
+        s_lengths: an array of single-byte values specifying the lengths of suffixes to be hashed in ascending order. 
+            For example, if *s_lengths==[2, 3]* and *cs == True*, the suffixes hashed for "spaCy" would be "yC" and "yCa".
+        ps_search_chars: a byte array containing, in numerical order, UTF-8 characters to search for within each token, 
+            starting at the beginning.
+        ps_width_offsets: an array of single-byte values [1-char-start, 2-char-start, 3-char-start, 4-char-start, 4-char-end]
+            specifying the offsets within *ps_search_chars* that contain UTF-8 characters with the specified widths.
+        ps_lengths: an array of single-byte values specifying the lengths of search results (from the beginning) to be hashed 
+            in ascending order. For example, if *ps_lengths==[1, 2]*, *ps_search=="aC" and *cs==False*, the searched strings 
+            hashed for "spaCy" would be "a" and "ac".
+        ss_search_chars: a byte array containing, in numerical order, UTF-8 characters to search for within each token, 
+            starting at the end.
+        ss_width_offsets: an array of single-byte values [1-char-start, 2-char-start, 3-char-start, 4-char-start, 4-char-end]
+            specifying the offsets within *ss_search_chars* that contain UTF-8 characters with the specified widths.
+        ss_lengths: an array of single-byte values specifying the lengths of search results (from the end) to be hashed
+             in ascending order. For example, if *ss_lengths==[1, 2]*, *ss_search=="aC" and *cs==False*, the searched strings 
+             hashed for "spaCy" would be "c" and "ca".
+    
+        Many of the buffers passed into and used by this method contain single-byte numerical values. This takes advantage of 
+        the fact that we are hashing short affixes and searching for small groups of characters. The calling code is responsible
+        for ensuring that lengths being passed in cannot exceed 63 and hence that resulting values with maximally four-byte 
+        character widths can never exceed 255.
+
+        Note that this method performs no data validation itself as it expects the calling code will already have done so, and
+        that the behaviour of the code may be erratic if the supplied parameters do not conform to expectations.
+        """
+        
+        # Work out lengths
+        cdef int p_lengths_l = strlen(<char*> p_lengths)
```

Comment:
We already have the lengths at the call site. Also, character pointers are not great. Can we use a typed memoryview for these parameters?

Reply:
I already tried both using a memoryview and passing in the lengths as parameters, and both slowed things down:

- the memoryview slowed things down massively (when I profiled the code, around 50% of the runtime was memoryview processing, and when I removed the memoryview the run time halved)
- passing in the lengths slowed things down slightly. Presumably passing them in from Python, and converting them from Python to C, takes longer than redetermining them within the C code?

I know a character pointer is somewhat counterintuitive here, but the calling code guarantees that each length occupies exactly one byte.

Reply:
> the memoryview slowed things down massively (when I profiled the code, around 50% of the runtime was memoryview processing, and when I removed the memoryview the run time halved)

Did you try to use the data pointer of the typed memory view? (It could be that the construction overhead is too large.)

Reply:
Creating a Cython memoryview from a Python `bytes` object seems to have a much greater overhead than simply using a pointer to it. I did a quick test passing a two-byte buffer `b"\2\3"` a million times from Python to the methods

```
def test_p(const unsigned char* char_p)
def test_v(const unsigned char[:] char_v)
```

both of which add the values of the two bytes together and return the results: the overall runtime was 3-4x longer using the memoryview than using the pointer. 

Eight separate buffers are passed into the method `get_character_combination_hashes()` in this PR; because the code is generally trimmed for effiency, it makes a big difference whether eight memoryviews have to be created or not.